kong/llm/init.lua

```diff
@@ -239,6 +270,104 @@ local function identify_request(request)
   end
 end
 
+local function get_token_text(event_t)
+  -- chat
+  return
+    event_t and
+    event_t.choices and
+    #event_t.choices > 0 and
+    event_t.choices[1].delta and
+    event_t.choices[1].delta.content
+
+    or
+
+  -- completions
+    event_t and
+    event_t.choices and
+    #event_t.choices > 0 and
+    event_t.choices[1].text
+
+    or ""
+end
+
+function _M:calculate_cost(query_body, tokens_models, tokens_factor, tokens_max_query_cost)
+  local query_cost = 0
+  local err
+
+  -- Check if max_tokens is provided in the request body
+  local max_tokens = query_body.max_tokens
+  
+  if not max_tokens then
+      if query_body.model and tokens_models then
+          max_tokens = tonumber(tokens_models[query_body.model])
+      end
+  end
+
+  if not max_tokens then
+      return nil, "No max_tokens in query and no key found in the plugin config for model: " .. query_body.model
+  end
+
+  -- Function to count the number of words in a string
+  local function count_words(str)
+      local count = 0
+      for word in str:gmatch("%S+") do
+          count = count + 1
+      end
+      return count
+  end
```

Comment:
Doesn't seem to capture any upvalues, so this could be moved to the outer scope.

Reply:
I don't get it

Reply:
Fixed in 4d3d97ad4a155ab7fbf06984e6e946ed408b15e7